README.md
[![Docker Pulls](https://img.shields.io/docker/pulls/panzer1119/mc-backup.svg)](https://hub.docker.com/r/panzer1119/mc-backup)
[![release](https://github.com/Panzer1119/docker-mc-backup/workflows/release/badge.svg?branch=multiarch)](https://github.com/Panzer1119/docker-mc-backup/actions?query=workflow%3Arelease)
[![Discord](https://img.shields.io/discord/660567679458869252?label=Discord&logo=discord)](https://discord.gg/DXfKpjB)

Provides a side-car container to backup itzg/minecraft-server world data.

## Environment variables

##### Common variables:

- `SRC_DIR`=/data
- `BACKUP_NAME`=world
- `INITIAL_DELAY`=2m
- `BACKUP_INTERVAL`=24h
- `PAUSE_IF_NO_PLAYERS`=false
- `PLAYERS_ONLINE_CHECK_INTERVAL`=5m
- `PRUNE_BACKUPS_DAYS`=7
- `PRUNE_RESTIC_RETENTION`=--keep-within 7d
- `SERVER_PORT`=25565
- `RCON_HOST`=localhost
- `RCON_PORT`=25575
- `RCON_PASSWORD`=minecraft
- `RCON_PASSWORD_FILE`: Can be set to read the RCON password from a file. Overrides `RCON_PASSWORD` if both are set.
- `RCON_RETRIES`=5 : Set to a negative value to retry indefinitely
- `RCON_RETRY_INTERVAL`=10s
- `EXCLUDES`=\*.jar,cache,logs
- `BACKUP_METHOD`=tar
- `RESTIC_ADDITIONAL_TAGS`=mc_backups
- `TZ` : Can be set to the timezone to use for logging

If `PRUNE_BACKUPS_DAYS` is set to a positive number, it'll delete old `.tgz` backup files from `DEST_DIR`. By default deletes backups older than a week.

If `BACKUP_INTERVAL` is set to 0 or smaller, script will run once and exit.

Both `INITIAL_DELAY` and `BACKUP_INTERVAL` accept times in `sleep` format: `NUMBER[SUFFIX] NUMBER[SUFFIX] ...`.
SUFFIX may be 's' for seconds (the default), 'm' for minutes, 'h' for hours or 'd' for days.

Examples:
- `BACKUP_INTERVAL`="1.5d" -> backup every one and a half days (36 hours)
- `BACKUP_INTERVAL`="2h 30m" -> backup every two and a half hours
- `INITIAL_DELAY`="120" -> wait 2 minutes before starting

The `PAUSE_IF_NO_PLAYERS` option lets you pause backups if no players are online.

If `PAUSE_IF_NO_PLAYERS`="true" and there are no players online after a backup is made, then instead of immediately scheduling the next backup, the script will start checking the server's player count every `PLAYERS_ONLINE_CHECK_INTERVAL` (defaults to 5 minutes). Once a player joins the server, the next backup will be scheduled in `BACKUP_INTERVAL`.

`EXCLUDES` is a comma-separated list of glob(3) patterns to exclude from backups. By default excludes all jar files (plugins, server files), logs folder and cache (used by i.e. PaperMC server).

##### `tar` backup method

- `DEST_DIR`=/backups
- `LINK_LATEST`=false
- `TAR_COMPRESS_METHOD`=gzip
- `ZSTD_PARAMETERS`=-3 --long=25 --single-thread

`LINK_LATEST` is a true/false flag that creates a symbolic link to the latest backup.

`TAR_COMPRESS_METHOD` is the compression method used by tar. Valid value: gzip bzip2 zstd

`ZSTD_PARAMETERS` sets the parameters for `zstd` compression. The `--long` parameter affects RAM requirements for both compression and decompression (the default of 25 means 2^25 bytes = 32 MB).

##### `restic` backup method

See [restic documentation](https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html) on what variables are needed to be defined.
At least one of `RESTIC_PASSWORD*` variables need to be defined, along with `RESTIC_REPOSITORY`.

Use the `RESTIC_ADDITIONAL_TAGS` variable to define a space separated list of additional restic tags. The backup will always be tagged with the value of `BACKUP_NAME`. e.g.: `RESTIC_ADDITIONAL_TAGS=mc_backups foo bar` will tag your backup with `foo`, `bar`, `mc_backups` and the value of `BACKUP_NAME`.

You can finetune the retention cycle of the restic backups using the `PRUNE_RESTIC_RETENTION` variable. Take a look at the [restic documentation](https://restic.readthedocs.io/en/latest/060_forget.html) for details.

> **_EXAMPLE_**  
> Setting `PRUNE_RESTIC_RETENTION` to `--keep-daily 7 --keep-weekly 5 --keep-monthly 12 --keep-yearly 75` will keep the most recent 7 daily snapshots, then 4 (remember, 7 dailies already include a week!) last-day-of-the-weeks and 11 or 12 last-day-of-the-months (11 or 12 depends if the 5 weeklies cross a month). And finally 75 last-day-of-the-year snapshots. All other snapshots are removed.

:warning: | When using restic as your backup method, make sure that you fix your container hostname to a constant value! Otherwise, each time a container restarts it'll use a different, random hostname which will cause it not to rotate your backups created by previous instances!
---|---

:warning: | When using restic, at least one of `HOSTNAME` or `BACKUP_NAME` must be unique, when sharing a repository. Otherwise other instances using the same repository might prune your backups prematurely.
---|---

:warning: | SFTP restic backend is not directly supported. Please use RCLONE backend with SFTP support.
---|---

##### `rclone` backup method
Rclone acts as the `tar` backup method but automatically moves the compressed files to a remote drive via [rclone](https://rclone.org/).

There are a few special environment variables for the rclone method.

- `RCLONE_REMOTE` is the name of the remote you've configured in your rclone.conf, see [remote setup](https://rclone.org/remote_setup/).
- `RCLONE_COMPRESS_METHOD`=gzip
- `DEST_DIR`=/backups is the container path where the archive is temporarily created
- `RCLONE_DEST_DIR` is the directory on the remote

Other parameters such as `PRUNE_BACKUPS_DAYS`, `ZSTD_PARAMETERS`, and `BACKUP_NAME` are all used as well.

**Note** that you will need to place your rclone config file in `/config/rclone/rclone.conf`.
This can be done by adding it through docker-compose,

```yaml
- ./rclone.config:/config/rclone/rclone.conf:ro
```
or by running the config wizard in a container and mounting the volume.
```shell
docker run -it --rm -v rclone-config:/config/rclone rclone/rclone config
```

then you must bind the volume **for the mc-backup process**
```yaml
volumes:
  - rclone-config:/config/rclone
```
**and the service**
```yaml
volumes:
  rclone-config:
    external: true
```

##### `borg` backup method
[Borg](https://www.borgbackup.org/) deduplicates and compresses the files and stores them in a repository which can be on the local host or a remote host.
See [borg documentation](https://borgbackup.readthedocs.io/) for more information.

Borg also supports encrypted repositories, but for now it's not implemented here, therefore it creates an unencrypted repository.

There are a few special environment variables for the borg method.

- `BORG_REPO`=/borg is the path or url to the [borg repository](https://borgbackup.readthedocs.io/en/stable/usage/general.html#repository-urls)
- `BORG_COMPRESS_METHOD`=lz4 [available compression methods](https://borgbackup.readthedocs.io/en/stable/internals/data-structures.html#compression)
- `BORG_ARCHIVE_PREFIX` is the prefix that gets prepended to the borg archvie name
- `BORG_ARCHIVE_SUFFIX` is the suffix that gets appended to the borg archvie name

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [commit-and-tag-version](https://github.com/absolute-version/commit-and-tag-version) for commit guidelines.

### [1.1.2](https://github.com/Panzer1119/docker-mc-backup/compare/v1.1.1...v1.1.2) (2022-08-07)


### Bug Fixes

* **borg:** add check if BORG_PRUNE_GFS is set and if not skip the parsing of it ([8f95916](https://github.com/Panzer1119/docker-mc-backup/commit/8f95916e4df4b17ac5415d7482b50cd62a7956da))
* **borg:** add default keep argument to prune command ([34870d1](https://github.com/Panzer1119/docker-mc-backup/commit/34870d1ffa15f161f1c4d17ecff384eaece91983))
* **borg:** missing space in value substring ([7f2080f](https://github.com/Panzer1119/docker-mc-backup/commit/7f2080f79965f58f7b49cc0baa07d61b8dc270fa))

### [1.1.1](https://github.com/Panzer1119/docker-mc-backup/compare/v1.1.0...v1.1.1) (2022-08-07)


### Bug Fixes

* **borg:** use mapfile to split command output instead of unquoted command expansion in an array ([5a3c8b3](https://github.com/Panzer1119/docker-mc-backup/commit/5a3c8b3c76ec99d4a139587353e25237a63334d6))

## [1.1.0](https://github.com/Panzer1119/docker-mc-backup/compare/v1.0.0...v1.1.0) (2022-08-07)


### Features

* **borg:** implement retention policy Grandfather-father-son ([efcfb37](https://github.com/Panzer1119/docker-mc-backup/commit/efcfb375bcb2f1e71d19d0ff4911b44a421c7fc4))


### Bug Fixes

* **borg:** prefix when pruning with borg ([9e10bd0](https://github.com/Panzer1119/docker-mc-backup/commit/9e10bd015fde84eb83c9ffe23c841c36ead2a724))

## 1.0.0 (2022-08-05)


### Features

* **borg:** create borg_common_options and borg_options in borgs init method ([bb4f6ac](https://github.com/Panzer1119/docker-mc-backup/commit/bb4f6accf958ffcf1b5723750bdd2f9541a7752e))
* **borg:** create new backup method borg ([3ccce0b](https://github.com/Panzer1119/docker-mc-backup/commit/3ccce0ba761d3367dbd4c238679f64995c2c019e))
* **borg:** implement new backup method borg ([fb911d9](https://github.com/Panzer1119/docker-mc-backup/commit/fb911d90b6c1c4ec03797a16af54948f163ca868))
* **borg:** set /borg as default value for env variable BORG_REPO ([a98068e](https://github.com/Panzer1119/docker-mc-backup/commit/a98068ef003a532567669b6e48d4708fb57fe86d))
* **borg:** use borg_common_options and borg_options in borgs backup and prune methods ([e4a3793](https://github.com/Panzer1119/docker-mc-backup/commit/e4a37938999441cb7ef8e2ba72e695af3da7ffaf))
* **restic:** add options to finetune restic ([#13](https://github.com/Panzer1119/docker-mc-backup/issues/13)) ([368a30e](https://github.com/Panzer1119/docker-mc-backup/commit/368a30e1247ab8aaf672b59ed5fd59c6eb6a745b))


### Bug Fixes

* **borg:** prefix when pruning with borg ([6e96ebd](https://github.com/Panzer1119/docker-mc-backup/commit/6e96ebdb5249d7f2a80446cd494ce82c20926a08))
* **borg:** use utc when generating timestamp for the archive name ([0ba5780](https://github.com/Panzer1119/docker-mc-backup/commit/0ba578060a2a28304c9e9853722171a283dc1918))
* ensured date in filenames honors TZ ([feeecee](https://github.com/Panzer1119/docker-mc-backup/commit/feeeceece1b4a3ca4307a00fa63dc4aef84c34c2)), closes [#42](https://github.com/Panzer1119/docker-mc-backup/issues/42)


backup-loop.sh
#!/bin/bash

set -euo pipefail

if [ "${DEBUG:-false}" == "true" ]; then
  set -x
fi

: "${SRC_DIR:=/data}"
: "${DEST_DIR:=/backups}"
: "${BACKUP_NAME:=world}"
: "${INITIAL_DELAY:=2m}"
: "${BACKUP_INTERVAL:=${INTERVAL_SEC:-24h}}"
: "${PAUSE_IF_NO_PLAYERS:=false}"
: "${PLAYERS_ONLINE_CHECK_INTERVAL:=5m}"
: "${BACKUP_METHOD:=tar}"        # currently one of tar, restic, borg
: "${TAR_COMPRESS_METHOD:=gzip}" # bzip2 gzip zstd
: "${ZSTD_PARAMETERS:=-3 --long=25 --single-thread}"
: "${PRUNE_BACKUPS_DAYS:=7}"
: "${PRUNE_RESTIC_RETENTION:=--keep-within ${PRUNE_BACKUP_DAYS:-7}d}"
: "${SERVER_PORT:=25565}"
: "${RCON_HOST:=localhost}"
: "${RCON_PORT:=25575}"
: "${RCON_PASSWORD_FILE:=}"
: "${RCON_PASSWORD:=minecraft}"
: "${RCON_RETRIES:=5}"
: "${RCON_RETRY_INTERVAL:=10s}"
: "${EXCLUDES:=*.jar,cache,logs}" # Comma separated list of glob(3) patterns
: "${LINK_LATEST:=false}"
: "${RESTIC_ADDITIONAL_TAGS:=mc_backups}" # Space separated list of restic tags
: "${XDG_CONFIG_HOME:=/config}"           # for rclone's base config path
: "${ONE_SHOT:=false}"
: "${TZ:=Etc/UTC}"
: "${RCLONE_COMPRESS_METHOD:=gzip}"
: "${RCLONE_REMOTE:=}"
: "${RCLONE_DEST_DIR:=}"
: "${BORG_REPO:=/borg}"
: "${BORG_COMPRESS_METHOD:=lz4}"
: "${BORG_ARCHIVE_PREFIX:=}"
: "${BORG_ARCHIVE_SUFFIX:=}"
: "${BORG_BASE_DIR:=/tmp/borg}"
: "${BORG_RELOCATED_REPO_ACCESS_IS_OK:=yes}"
: "${BORG_UNKNOWN_UNENCRYPTED_REPO_ACCESS_IS_OK:=yes}"
: "${BORG_PRUNE_GFS:=}"
: "${VERBOSE:=false}"
export TZ

export RCON_HOST
export RCON_PORT
export RCON_PASSWORD
export XDG_CONFIG_HOME
export BORG_BASE_DIR
export BORG_RELOCATED_REPO_ACCESS_IS_OK
export BORG_UNKNOWN_UNENCRYPTED_REPO_ACCESS_IS_OK

###############
##  common   ##
## functions ##
###############

is_one_shot() {
  if [[ ${ONE_SHOT^^} = TRUE ]]; then
    return 0
  else
    return 1
  fi
}

is_elem_in_array() {
  # $1 = element
  # All remaining arguments are array to search for the element in
  if [ "$#" -lt 2 ]; then
    log INTERNALERROR "Wrong number of arguments passed to is_elem_in_array function"
    return 2
  fi
  local element="${1}"
  shift
  local e
  for e; do
    if [ "${element}" == "${e}" ]; then
      return 0
    fi
  done
  return 1
}

log() {
  if [ "$#" -lt 1 ]; then
    log INTERNALERROR "Wrong number of arguments passed to log function"
    return 2
  fi
  local level="${1}"
  shift
  local valid_levels=(
    "INFO"
    "WARN"
    "ERROR"
    "INTERNALERROR"
  )
  if ! is_elem_in_array "${level}" "${valid_levels[@]}"; then
    log INTERNALERROR "Log level ${level} is not a valid level."
    return 2
  fi
  (
    # If any arguments are passed besides log level
    if [ "$#" -ge 1 ]; then
      # then use them as log message(s)
      cat <<<"${*}" -
    else
      # otherwise read log messages from standard input
      cat -
    fi
    if [ "${level}" == "INTERNALERROR" ]; then
      echo "Please report this: https://github.com/Panzer1119/docker-mc-backup/issues"
    fi
  ) | awk -v level="${level}" '{ printf("%s %s %s\n", strftime("%FT%T%z"), level, $0); fflush(); }'
} >&2

retry() {
  if [ "$#" -lt 3 ]; then
    log INTERNALERROR "Wrong number of arguments passed to retry function"
    return 1
  fi

  # How many times should we retry?
  # Value smaller than zero means infinitely
  local retries="${1}"
  # Time to sleep between retries
  local interval="${2}"
  readonly retries interval
  shift 2

  if ((retries < 0)); then
    local retries_msg="infinite"
  else
    local retries_msg="${retries}"
  fi

  local i=-1 # -1 since we will increment it before printing
  while ((retries >= ++i)) || [ "${retries_msg}" != "${retries}" ]; do
    # Send SIGINT after 5 minutes. If it doesn't shut down in 30 seconds, kill it.
    if output="$(timeout --signal=SIGINT --kill-after=30s 5m "${@}" 2>&1 | tr '\n' '\t')"; then
      log INFO "Command executed successfully ${*}"
      return 0
    else
      log ERROR "Unable to execute ${*} - try ${i}/${retries_msg}. Retrying in ${interval}"
      if [ -n "${output}" ]; then
        log ERROR "Failure reason: ${output}"
      fi
    fi
    # shellcheck disable=SC2086
    sleep ${interval}
  done
  return 2
}

is_function() {
  if [ "${#}" -ne 1 ]; then
    log INTERNALERROR "is_function expects 1 argument, received ${#}"
  fi
  name="${1}"
  [ "$(type -t "${name}")" == "function" ]
}

call_if_function_exists() {
  if [ "${#}" -lt 1 ]; then
    log INTERNALERROR "call_if_function_exists expects at least 1 argument, received ${#}"
    return 2
  fi
  function_name="${1}"
  if is_function "${function_name}"; then
    "${@}"
  else
    log INTERNALERROR "${function_name} is not a valid function!"
    return 2
  fi
}

#####################
## specific method ##
##    functions    ##
#####################
# Each function that corresponds to a name of a backup method
# Should define following functions inside them
# init() -> called before entering loop. Verify arguments, prepare for operations etc.
# backup() -> create backup. It's guaranteed that all data is already flushed to disk.
# prune() -> prune old backups. PRUNE_BACKUPS_DAYS is guaranteed to be positive.

tar() {
  _find_old_backups() {
    find "${DEST_DIR}" -maxdepth 1 -name "*.${backup_extension}" -mtime "+${PRUNE_BACKUPS_DAYS}" "${@}"
  }

  init() {
    mkdir -p "${DEST_DIR}"
    case "${TAR_COMPRESS_METHOD}" in
    gzip)
      readonly tar_parameters=("--gzip")
      readonly backup_extension="tgz"
      ;;

    bzip2)
      readonly tar_parameters=("--bzip2")
      readonly backup_extension="bz2"
      ;;

    zstd)
      readonly tar_parameters=("--use-compress-program" "zstd ${ZSTD_PARAMETERS}")
      readonly backup_extension="tar.zst"
      ;;

    *)
      log ERROR 'TAR_COMPRESS_METHOD is not valid!'
      exit 1
      ;;
    esac
  }
  backup() {
    ts=$(date +"%Y%m%d-%H%M%S")
    outFile="${DEST_DIR}/${BACKUP_NAME}-${ts}.${backup_extension}"
    tries=3
    while ((tries-- > 0)); do
      log INFO "Backing up content in ${SRC_DIR} to ${outFile}"
      command tar "${excludes[@]}" "${tar_parameters[@]}" -cf "${outFile}" -C "${SRC_DIR}" . || exitCode=$?
      if [ ${exitCode:-0} -eq 0 ]; then
        break
      elif [ ${exitCode:-0} -eq 1 ]; then
        if ((tries > 0)); then
          log INFO "...retrying backup in 5 seconds"
          sleep 5
          continue
        else
          log WARN "Giving up on this round of backup"
        fi
      elif [ ${exitCode:-0} -gt 1 ]; then
        log ERROR "tar exited with code ${exitCode}! Aborting"
        exit 1
      fi
    done
    if [ "${LINK_LATEST^^}" == "TRUE" ]; then
      ln -sf "${BACKUP_NAME}-${ts}.${backup_extension}" "${DEST_DIR}/latest.${backup_extension}"
    fi
  }
  prune() {
    if [ -n "$(_find_old_backups -print -quit)" ]; then
      log INFO "Pruning backup files older than ${PRUNE_BACKUPS_DAYS} days"
      _find_old_backups -print -delete | awk '{ printf "Removing %s\n", $0 }' | log INFO
    fi
  }
  call_if_function_exists "${@}"
}

restic() {
  _delete_old_backups() {
    # shellcheck disable=SC2086
    command restic forget --tag "${restic_tags_filter}" ${PRUNE_RESTIC_RETENTION} "${@}"
  }
  _check() {
    if ! output="$(command restic check 2>&1)"; then
      log ERROR "Repository contains error! Aborting"
      log <<<"${output}" ERROR
      return 1
    fi
  }
  init() {
    if [ -z "${RESTIC_PASSWORD:-}" ] &&
      [ -z "${RESTIC_PASSWORD_FILE:-}" ] &&
      [ -z "${RESTIC_PASSWORD_COMMAND:-}" ]; then
      log ERROR "At least one of" RESTIC_PASSWORD{,_FILE,_COMMAND} "needs to be set!"
      return 1
    fi
    if [ -z "${RESTIC_REPOSITORY:-}" ]; then
      log ERROR "RESTIC_REPOSITORY is not set!"
      return 1
    fi
    if output="$(command restic snapshots 2>&1 >/dev/null)"; then
      log INFO "Repository already initialized"
      _check
    elif grep <<<"${output}" -q '^Is there a repository at the following location?$'; then
      log INFO "Initializing new restic repository..."
      command restic init | log INFO
    elif grep <<<"${output}" -q 'wrong password'; then
      log <<<"${output}" ERROR
      log ERROR "Wrong password provided to an existing repository?"
      return 1
    else
      log <<<"${output}" ERROR
      log INTERNALERROR "Unhandled restic repository state."
      return 2
    fi

    # Used to construct tagging arguments and filters for snapshots
    read -ra restic_tags <<<${RESTIC_ADDITIONAL_TAGS}
    restic_tags+=("${BACKUP_NAME}")
    readonly restic_tags

    # Arguments to use to tag the snapshots with
    restic_tags_arguments=()
    local tag
    for tag in "${restic_tags[@]}"; do
      restic_tags_arguments+=(--tag "$tag")
    done
    readonly restic_tags_arguments
    # Used for filtering backups to only match ours
    restic_tags_filter="$(
      IFS=,
      echo "${restic_tags[*]}"
    )"
    readonly restic_tags_filter
  }
  backup() {
    log INFO "Backing up content in ${SRC_DIR}"
    command restic backup "${restic_tags_arguments[@]}" "${excludes[@]}" "${SRC_DIR}" | log INFO
  }
  prune() {
    # We cannot use `grep -q` here - see https://github.com/restic/restic/issues/1466
    if _delete_old_backups --dry-run | grep '^remove [[:digit:]]* snapshots:$' >/dev/null; then
      log INFO "Pruning snapshots using ${PRUNE_RESTIC_RETENTION}"
      _delete_old_backups --prune | log INFO
      _check | log INFO
    fi
  }
  call_if_function_exists "${@}"
}

rclone() {
  _find_old_backups() {
    command rclone lsf --format "tp" "${RCLONE_REMOTE}:${RCLONE_DEST_DIR}" | grep ${BACKUP_NAME} | awk \
      -v PRUNE_DATE="$(date '+%Y-%m-%d %H:%M:%S' --date="${PRUNE_BACKUPS_DAYS} days ago")" \
      -v DESTINATION="${RCLONE_DEST_DIR%/}" \
      'BEGIN { FS=";" } $1 < PRUNE_DATE { printf "%s/%s\n", DESTINATION, $2 }'
  }
  init() {
    # Check if rclone is installed and configured correctly
    mkdir -p "${DEST_DIR}"
    case "${RCLONE_COMPRESS_METHOD}" in
    gzip)
      readonly tar_parameters=("--gzip")
      readonly backup_extension="tgz"
      ;;

    bzip2)
      readonly tar_parameters=("--bzip2")
      readonly backup_extension="bz2"
      ;;

    zstd)
      readonly tar_parameters=("--use-compress-program" "zstd ${ZSTD_PARAMETERS}")
      readonly backup_extension="tar.zst"
      ;;

    *)
      log ERROR 'RCLONE_COMPRESS_METHOD is not valid!'
      exit 1
      ;;
    esac
  }
  backup() {
    ts=$(date +"%Y%m%d-%H%M%S")
    outFile="${DEST_DIR}/${BACKUP_NAME}-${ts}.${backup_extension}"
    log INFO "Backing up content in ${SRC_DIR} to ${outFile}"
    command tar "${excludes[@]}" "${tar_parameters[@]}" -cf "${outFile}" -C "${SRC_DIR}" . || exitCode=$?
    if [ ${exitCode:-0} -eq 1 ]; then
      log WARN "tar exited with code 1. Ignoring"
    fi
    if [ ${exitCode:-0} -gt 1 ]; then
      log ERROR "tar exited with code ${exitCode}! Aborting"
      exit 1
    fi

    command rclone copy "${outFile}" "${RCLONE_REMOTE}:${RCLONE_DEST_DIR}"
    rm "${outFile}"
  }
  prune() {
    if [ -n "$(_find_old_backups)" ]; then
      log INFO "Pruning backup files older than ${PRUNE_BACKUPS_DAYS} days"
      _find_old_backups | tee \
        >(awk '{ printf "Removing %s\n", $0 }' | log INFO) \
        >(while read -r path; do command rclone deletefile "${RCLONE_REMOTE}:${path}"; done)
    fi
  }
  call_if_function_exists "${@}"
}

borg() {
  _check() {
    if ! output="$(command borg check "${BORG_REPO}" 2>&1)"; then
      log ERROR "Borg repository contains errors! Aborting"
      log <<<"${output}" ERROR
      return 1
    fi
  }
  init() {
    #    if [ -z "${RESTIC_PASSWORD:-}" ] \
    #        && [ -z "${RESTIC_PASSWORD_FILE:-}" ] \
    #        && [ -z "${RESTIC_PASSWORD_COMMAND:-}" ]; then
    #      log ERROR "At least one of" RESTIC_PASSWORD{,_FILE,_COMMAND} "needs to be set!"
    #      return 1
    #    fi
    if [ -z "${BORG_REPO:-}" ]; then
      log ERROR "BORG_REPO is not set!"
      return 1
    fi
    borg_common_options=()
    borg_options_create=()
    borg_options_prune=()
    if [ "${DEBUG:-false}" == "true" ]; then
      borg_common_options+=(--debug)
    fi
    if [ "${VERBOSE:-false}" == "true" ]; then
      borg_common_options+=(--verbose)
      borg_common_options+=(--progress)
      borg_options_create+=(--stats)
      borg_options_prune+=(--stats)
    fi
    readonly borg_common_options
    readonly borg_options_create
    if output="$(command borg info "${BORG_REPO}" 2>&1 >/dev/null)"; then
      log INFO "Borg repository already initialized"
      _check
    else
      log INFO "Initializing new borg repository..."
      command borg "${borg_common_options[@]}" init --encryption none --make-parent-dirs "${BORG_REPO}"
      #TODO Encryption?
    fi
    borg_use_gfs=false
    #Borg GFS Parsing
    #borg_keep_secondly=-1 # Do not use this (then we don't need to decide whether to use s or S)
    borg_keep_yearly=-1
    borg_keep_monthly=-1
    borg_keep_weekly=-1
    borg_keep_daily=-1
    borg_keep_hourly=-1
    borg_keep_minutely=-1
    borg_keep_gfs_log=()
    if [ -n "${BORG_PRUNE_GFS}" ]; then
      mapfile -t borg_prune_gfs_array < <(echo "${BORG_PRUNE_GFS}" | tr "," "\n")
      local unit
      local value
      for i in "${borg_prune_gfs_array[@]}"; do
        unit="${i: -1}"
        value="${i::-1}"
        #log INFO "$unit: $value"
        case "${unit}" in

        y)
          #log INFO "Yearly"
          borg_keep_yearly="${value}"
          ;;

        m)
          #log INFO "Monthly"
          borg_keep_monthly="${value}"
          ;;

        w)
          #log INFO "weekly"
          borg_keep_weekly="${value}"
          ;;

        d)
          #log INFO "daily"
          borg_keep_daily="${value}"
          ;;

        H)
          #log INFO "hourly"
          borg_keep_hourly="${value}"
          ;;

        M)
          #log INFO "minutely"
          borg_keep_minutely="${value}"
          ;;

        *)
          log ERROR "Unknown unit \"${unit}\" in BORG_PRUNE_GFS"
          return 1
          ;;
        esac
      done
      if ((borg_keep_minutely > -1)); then
        borg_options_prune+=(--keep-minutely "${borg_keep_minutely}")
        borg_keep_gfs_log+=("${borg_keep_minutely} minutely")
        borg_use_gfs=true
      fi
      if ((borg_keep_hourly > -1)); then
        borg_options_prune+=(--keep-hourly "${borg_keep_hourly}")
        borg_keep_gfs_log+=("${borg_keep_hourly} hourly")
        borg_use_gfs=true
      fi
      if ((borg_keep_daily > -1)); then
        borg_options_prune+=(--keep-daily "${borg_keep_daily}")
        borg_keep_gfs_log+=("${borg_keep_daily} daily")
        borg_use_gfs=true
      fi
      if ((borg_keep_weekly > -1)); then
        borg_options_prune+=(--keep-weekly "${borg_keep_weekly}")
        borg_keep_gfs_log+=("${borg_keep_weekly} weekly")
        borg_use_gfs=true
      fi
      if ((borg_keep_monthly > -1)); then
        borg_options_prune+=(--keep-monthly "${borg_keep_monthly}")
        borg_keep_gfs_log+=("${borg_keep_monthly} monthly")
        borg_use_gfs=true
      fi
      if ((borg_keep_yearly > -1)); then
        borg_options_prune+=(--keep-yearly "${borg_keep_yearly}")
        borg_keep_gfs_log+=("${borg_keep_yearly} yearly")
        borg_use_gfs=true
      fi
    fi
    if [ "${borg_use_gfs}" == "false" ]; then
      borg_options_prune+=(--keep-within "${PRUNE_BACKUPS_DAYS}d")
    fi
    readonly borg_options_prune
    #log INFO "borg_options_prune: ${borg_options_prune[*]}"

    #    if output="$(command restic snapshots 2>&1 >/dev/null)"; then
    #      log INFO "Repository already initialized"
    #      _check
    #    elif <<<"${output}" grep -q '^Is there a repository at the following location?$'; then
    #      log INFO "Initializing new restic repository..."
    #      command restic init | log INFO
    #    elif <<<"${output}" grep -q 'wrong password'; then
    #      <<<"${output}" log ERROR
    #      log ERROR "Wrong password provided to an existing repository?"
    #      return 1
    #    else
    #      <<<"${output}" log ERROR
    #      log INTERNALERROR "Unhandled restic repository state."
    #      return 2
    #    fi
  }
  backup() {
    local ts
    local cwd
    local archive
    ts=$(date --utc +"%Y%m%d-%H%M%S")
    #ts=$(date --utc --iso-8601=seconds) # Nicer ISO 8601 Format
    cwd=$(pwd)
    archive="${BORG_ARCHIVE_PREFIX:=}${BACKUP_NAME}-${ts}${BORG_ARCHIVE_SUFFIX:=}"
    log INFO "Backing up content in ${SRC_DIR} to ${BORG_REPO}::${archive}"
    cd "${SRC_DIR}"
    command borg "${borg_common_options[@]}" create "${borg_options_create[@]}" --numeric-ids --compression "${BORG_COMPRESS_METHOD:=lz4}" "${excludes[@]}" "${BORG_REPO}"::"${archive}" . | log INFO
    cd "${cwd}"
  }
  prune() {
    if [ "${borg_use_gfs}" == "false" ]; then
      log INFO "Pruning borg archives older than ${PRUNE_BACKUPS_DAYS} days"
    else
      local joined
      for i in "${borg_keep_gfs_log[@]}"; do
        joined+="${i}, "
      done
      log INFO "Pruning borg archives, keeping ${joined::-2} archives"
    fi
    command borg "${borg_common_options[@]}" prune "${borg_options_prune[@]}" --prefix "${BORG_ARCHIVE_PREFIX:=}${BACKUP_NAME}-" "${BORG_REPO}"
    command borg "${borg_common_options[@]}" compact "${BORG_REPO}"
  }
  call_if_function_exists "${@}"
}

##########
## main ##
##########

if [[ $RCON_PASSWORD_FILE ]]; then
  if [ ! -e ${RCON_PASSWORD_FILE} ]; then
    log ERROR "Initial RCON password file ${RCON_PASSWORD_FILE} does not seems to exist."
    log ERROR "Please ensure your configuration."
    log ERROR "If you are using Docker Secrets feature, please check this for further information: "
    log ERROR " https://docs.docker.com/engine/swarm/secrets"
    exit 1
  else
    RCON_PASSWORD=$(cat ${RCON_PASSWORD_FILE})
    export RCON_PASSWORD
  fi
fi

if [ -n "${INTERVAL_SEC:-}" ]; then
  log WARN 'INTERVAL_SEC is deprecated. Use BACKUP_INTERVAL instead'
fi

if [ ! -d "${SRC_DIR}" ]; then
  log ERROR 'SRC_DIR does not point to an existing directory!'
  exit 1
fi

if ! is_function "${BACKUP_METHOD}"; then
  log ERROR "Invalid BACKUP_METHOD provided: ${BACKUP_METHOD}"
fi

# We unfortunately can't use a here-string, as it inserts new line at the end
readarray -td, excludes_patterns < <(printf '%s' "${EXCLUDES}")

excludes=()
for pattern in "${excludes_patterns[@]}"; do
  excludes+=(--exclude "${pattern}")
done

"${BACKUP_METHOD}" init

if ! is_one_shot; then
  log INFO "waiting initial delay of ${INITIAL_DELAY}..."
  # shellcheck disable=SC2086
  sleep ${INITIAL_DELAY}
fi

log INFO "waiting for rcon readiness..."
retry ${RCON_RETRIES} ${RCON_RETRY_INTERVAL} rcon-cli save-on

while true; do

  if retry ${RCON_RETRIES} ${RCON_RETRY_INTERVAL} rcon-cli save-off; then
    # No matter what we were doing, from now on if the script crashes
    # or gets shut down, we want to make sure saving is on
    trap 'retry 5 5s rcon-cli save-on' EXIT

    retry ${RCON_RETRIES} ${RCON_RETRY_INTERVAL} rcon-cli save-all flush
    retry ${RCON_RETRIES} ${RCON_RETRY_INTERVAL} sync

    "${BACKUP_METHOD}" backup

    retry ${RCON_RETRIES} ${RCON_RETRY_INTERVAL} rcon-cli save-on
    # Remove our exit trap now
    trap EXIT
  else
    log ERROR "Unable to turn saving off. Is the server running?"
    exit 1
  fi

  if ((PRUNE_BACKUPS_DAYS > 0)); then
    "${BACKUP_METHOD}" prune
  fi

  if is_one_shot; then
    break
  fi

  # If BACKUP_INTERVAL is not a valid number (i.e. 24h), we want to sleep.
  # Only raw numeric value <= 0 will break
  if ((BACKUP_INTERVAL <= 0)) &>/dev/null; then
    break
  fi

  if [[ ${PAUSE_IF_NO_PLAYERS^^} = TRUE ]]; then
    while true; do
      if ! PLAYERS_ONLINE=$(mc-monitor status --host "${RCON_HOST}" --port "${SERVER_PORT}" --show-player-count 2>&1); then
        log ERROR "Error querying the server, waiting ${PLAYERS_ONLINE_CHECK_INTERVAL}..."
        sleep "${PLAYERS_ONLINE_CHECK_INTERVAL}"
      elif [ "${PLAYERS_ONLINE}" = 0 ]; then
        log INFO "No players online, waiting ${PLAYERS_ONLINE_CHECK_INTERVAL}..."
        sleep "${PLAYERS_ONLINE_CHECK_INTERVAL}"
      else
        break
      fi
    done
  fi

  log INFO "sleeping ${BACKUP_INTERVAL}..."
  # shellcheck disable=SC2086
  sleep ${BACKUP_INTERVAL}
done
